Issue/7696 fix agent cache part 2 remove version based caching

Hugo-Inmanta opened the pull request:
# Description

Part 2 of https://github.com/inmanta/inmanta-core/issues/7696

- convert version-based caching to time based: each version is kept open for 60s after the last use of an item belonging to it
- remove mechanisms for manual handling of opening/closing a given version

- refactored `test_decorator`: split it in multiple test cases

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

arnaudsjs commented on tests/test_cache.py:
```diff
@@ -354,30 +275,35 @@ def __call__(self, **kwargs):
 
 
-def test_decorator():
+async def test_decorator(agent_cache):
```
I don't think we need the `agent_cache` fixture here. We can just call into `AgentCache()`. The latter will be faster as it doesn't start a server and an agent.

sanderr reviewed (commented):
I haven't progressed far yet because I had some major difficulties grasping the motivation. I'm releasing my comments now mainly because I've got one very big one related to that and it may spark some discussion.

Hugo-Inmanta commented on src/inmanta/agent/cache.py:
```diff
@@ -63,51 +62,33 @@ def __init__(
-    2. version
+    cache items are removed from the cache either:
+    1. individually when their expiry_time is up
+    2. as a group when their version is not being used for a while (default 1 min)
```
1. Agree
2. 
> for which for_version=True was set  

I think `for_version` is always implicitly set to True, maybe more accurate would be "for which version=x was set"?
> we want to expire phased out versions' cache entries even if their lifetime has not ended yet, in order to keep the memory footprint acceptable

I agree
> And "last used" refers to the version being used by the agent (cache.open_version() or similar), **rather than any of its cache entries being used**

Why not ? In my mind these are equivalent (any of its cache entries _for this version_ being used)

sanderr replied:
> always implicitly set to True

Good catch, I forgot to check and just assumed the default was `False`. That revelation makes this whole thing more relevant, though it still all depends on whether the user passes the version or not.

We could consider to change the default, because I can imagine its primary purpose used to be to allow new code to recalculate the value, but our executors already achieve that now (not entirely sure of the in-process ones, but I think even there each executor has their own cache?).

Or we can even go as far as to drop the `for_version` argument entirely, which might be what was meant with the proposal in the ticket (on the other hand, then the version-based timer suggestion wouldn't make sense anymore).

> Why not ? In my mind these are equivalent (any of its cache entries for this version being used)

I would think we'd want to keep a version's cache alive as long as the version is active, even if none of the resources that were deployed in the past minute happened to read from the cache. e.g. imagine an agent with a few resources of type `X` and many of type `Y`. The ones of type `X` use the cache. They are deployed in order `x0 y00 y01 ... x1 y10 y11 ...`. Suppose the deploy of `y00 .. y0n` takes more than a minute, I would think we'd want to keep the cache for `X` alive, because we're still actively deploying the version, and `x1` will need it.

sanderr replied:
I find it difficult to decide on the best approach. Its tough to trust to our judgement when it's clear we don't fully understand what exactly Wouter suggested and why. I'll get back to this.

Hugo-Inmanta replied:
> I would think we'd want to keep the cache for X alive, because we're still actively deploying the version, and x1 will need it

I agree

sanderr replied:
I discussed this with Wouter earlier today. Key takeaways:
- the general reasoning above is correct.
- keep the default for `for_version`
- for cache entries with `for_version=False`, use only the expiry time (time since creation)
- for cache entries with `for_version=True`, add a linger time: time since last use. No need to keep track of this per version, also no need to keep track of this since agent active. It suffices to track last access of the cache value.
- The cache should only expire while the agent is inactive. e.g. before each deploy (or other resource action) starts, freeze the cache, afterwards unfreeze it. While unfrozen, the periodic check is active. When frozen, the check is run once more and then paused until unfrozen. This ensures consistent cache entries for at least a single resource deploy. The deploy will then update any "last used" timers for cache entries it consumes.

@wouterdb is that the gist of it?

wouterdb replied:
sounds good to me

Hugo-Inmanta commented on src/inmanta/agent/cache.py:
```diff
@@ -63,51 +62,33 @@ def __init__(
-        self.time: float = time.time() + scope.timeout
         self.call_on_delete = call_on_delete
+        self.expiry_time = time.time() + scope.timeout
+        self.called_finalizer = False
```
I added `self.called_finalizer` with the intention to make sure we don't call it twice because the following test was failing without it, but I very well might be missing something:

https://github.com/inmanta/inmanta-core/blob/b6146e6f79c061fb533fdeb79d9621c89740893d/tests/test_cache.py#L244

```
>       assert alpha.deleted == alpha.created
E       assert 2 == 1
E        +  where 2 = <test_cache.test_multi_threaded.<locals>.Spy object at 0x76113933c4d0>.deleted
E        +  and   1 = <test_cache.test_multi_threaded.<locals>.Spy object at 0x76113933c4d0>.created
```

sanderr replied:
As far as I can tell the current implementation would indeed fix it in most cases, but since you have a non-thread safe check and write to this attribute, I think a race condition is possible when two threads call the method at roughly the same time.

That is, unless we are confident that that would never happen. If it's really about garbage collection and *one* explicit call, we can perhaps assume that garbage collection will only take place after we've finished all our calls, because we would still have a reference to the object while calling delete. But if we rely on that I would very clearly state why the variable is there and why it's safe without thread-safe access.

Hugo-Inmanta replied:
I added a lock to make read/write to `called_finalizer` thread-safe

sanderr reviewed (commented):
Postponing further review until https://github.com/inmanta/inmanta-core/pull/7850#discussion_r1704656836 has been addressed.

Hugo-Inmanta commented on src/inmanta/agent/cache.py:
```diff
@@ -132,70 +139,37 @@ def __init__(self, agent_instance: Optional["AgentInstance"] = None) -> None:
+
+        self.lingering_set = set()
+        if self.timer_queue:
+            # Bad O(N) but unavoidable (?) Since we modify elements above
+            # we have to make sure the heap is still a heap.
+            heapq.heapify(self.timer_queue)
+            self.next_action = self.timer_queue[0].expiry_time
```
I think this is pretty bad but I don't see a way around it.

wouterdb replied:
N is likely to be small, so I would leave it at this.

Hugo-Inmanta commented:
@arnaudsjs Re-requesting your review since this changed a lot since your last review. Main differences:
- refactored test_decorator: split it in multiple test cases
- 2 types of cache entries: lingering vs non-lingering
     - lingering entries stay for 60s after their last **use**
     - non-lingering expire `<timeout>` seconds after **entering** the cache

arnaudsjs commented on tests/test_cache.py:
```diff
@@ -373,135 +294,438 @@ def __call__(self, **kwargs):
-    class Closeable:
-        def __init__(self):
-            self.closed = False
+class CacheDecoratorTest:
+    def __init__(self, cache: AgentCache):
+        self.cache = cache
+        self.cache_miss_counter = 0
```
The name indicates this is a decorator, but it's not.

arnaudsjs commented on tests/test_cache.py:
```diff
@@ -373,135 +294,438 @@ def __call__(self, **kwargs):
+        agent_cache.close()
+        assert my_closable.closed
+
+        test.cache_miss_counter = 0
+        my_closable.closed = False
+
+        test.cache_miss_counter = 0
```
```suggestion
        test.cache_miss_counter = 0
        my_closable.closed = False
```

Hugo-Inmanta commented on docs/model_developers/handlers.rst:
```diff
@@ -237,28 +235,38 @@ the method will form the cache key, the return value will be cached. When the me
 .. code-block:: python
 
-    @cache(timeout=120, ignore=["ctx"])
+    @cache(timeout=120, ignore=["ctx"], for_version=False)
```
> Any opinion on the above-mentioned approach?

It seems perfect to me. What to do with this parameter had been bothering me, I didn't know how to rename it while also keeping backwards compatibility. Your approach checks all the boxes :+1: 

Follow-up ticket [here](https://github.com/inmanta/inmanta-core/issues/8052)

Hugo-Inmanta commented on src/inmanta/agent/cache.py:
```diff
@@ -224,69 +201,70 @@ def _get(self, key: str) -> CacheItem:
-        version: int = 0,
         timeout: int = 5000,
         call_on_delete: Optional[Callable[[Any], None]] = None,
+        for_version: bool = True,
```
Created a follow-up ticket as suggested https://github.com/inmanta/inmanta-core/pull/7850#discussion_r1745489297

Hugo-Inmanta commented on src/inmanta/agent/handler.py:
```diff
@@ -122,12 +122,15 @@ def cache(
 
     :param timeout: the number of second this cache entry should live
-    :param for_version: if true, this value is evicted from the cache when this deploy is ready
+    :param for_version: if True (default behaviour), the value being cached will be tied to a given version
+        passed via the ``version`` argument of the decorated method.
```
Nice catch

Hugo-Inmanta commented on src/inmanta/agent/cache.py:
```diff
@@ -38,76 +38,82 @@ class Scope:
-    """
-    A context manager to ensure the cache version is properly closed
+@stable_api
+class AgentCache:
```
> Do you know why the add_lock and add_locks are required? 

I'm not sure. I think these locks are redundant now that we have the `activity_lock` around all cache accesses. @wouterdb could you confirm ?

wouterdb replied:
the add_lock protects the dict of add_locks. Each add_lock protects a cache entry from being created twice at the same time. 

The mechanism is now also available in the utils as NamedLock
